modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,209 @@
+        # print("-----------------------------------------------")
+
+        # Sort weights from largest to smallest, along with corresponding covariances and means
+        indices = np.argsort(weights)[::-1]
+        weights = np.sort(weights)[::-1]
+        temp_arr = np.array(covariances)
+        for i in range(len(indices)):
+            covariances[i] = temp_arr[indices[i]]
+        temp_arr = np.array(clusters)
+        for i in range(len(indices)):
+            clusters[i] = temp_arr[indices[i]]
+
+        # print("Sorted:")
+        # print(weights)
+        # print(covariances)
+        # print(clusters)
+        # print("-----------------------------------------------")
+
+        # Loop through each cluster
+        # clusters is a list of centers ordered by weights
+        # most likely cluster listed first in descending weights order
+        objects_in_world = []
+        total_clusters = clusters.shape[0]
+        num_viable_clusters = 1
+        # Drop all clusters after a weight_drop_threshold drop in weight occured
+        while num_viable_clusters < total_clusters:
+            if (weights[num_viable_clusters-1] / weights[num_viable_clusters]) > (1 / self.__weight_drop_threshold):
+                weights = weights[:num_viable_clusters]
+                clusters = clusters[:num_viable_clusters]
+                covariances = covariances[:num_viable_clusters]
+                break
+            num_viable_clusters += 1
+
+        # print("Weight-filtered:")
+        # print(weights)
+        # print(covariances)
+        # print(clusters)
+        # print("-----------------------------------------------")
+
+        # Bad detection removal
+        i = 0
+        while i < len(clusters):
+            nearby_point = False
+            j = 0
+            while j < len(self.__all_points):
+                # Remove cluster if nothing within 2 metres
+                if self.get_distance(clusters[i], self.__all_points[j]) < 10:
+                    nearby_point = True
+                    break
+                j += 1
+            if not nearby_point:
+                clusters = np.delete(clusters, i, 0)
+                covariances = np.delete(covariances, i, 0)
+            else:
+                i += 1
+
+
+        print("Bad clusters removed:")
+        print(weights)
+        print(covariances)
+        print(clusters)
+        print("-----------------------------------------------")
+        print()
+        print()
+
+        # Create output list of remaining (valid) clusters 
+        for i in range(len(clusters)):
+            object_created, object_to_add = ObjectInWorld.create(
+                clusters[i][0], 
+                clusters[i][1], 
+                covariances[i]
+                )
+            if object_created:
+                objects_in_world.append(object_to_add)
+
+
+
+        return True, objects_in_world
+
+    def decide_to_run(self, detections: "list[DetectionInWorld]") -> bool:
```

Comment:
Is this called outside of the class? If not, make private by appending 2 underscores.